SERVER-81279 Create Genny Workloads for UpdateMany

brettnawrocki opened the pull request:
**Jira Ticket:** [SERVER-81279](https://jira.mongodb.org/browse/SERVER-81279)

**Whats Changed:** 
Adds performance workloads for various cases involving updateMany in sharded clusters.

**Patch testing results:**  
https://spruce.mongodb.com/version/6597130356234348f74ef21a/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

dalyd commented:
This passed through my inbox, but is not assigned to me. So two quick thoughts:
I think you probably want product-performance instead of performance-infrastructure on this. @mongodb/performance 

Also, could you add Keywords along with description and owner?

brettnawrocki commented:
@dalyd Sorry for the mixup! I read somewhere to add performance infrastructure for genny PRs, but in retrospect I suppose that means core genny code and not genny workloads. 😅 

I can add keywords to the workloads as well. Thanks for taking a glace.

thessem commented:
Sorry about the confusion @brettnawrocki, I've got an eye on using [DEVPROD-3142](https://jira.mongodb.org/browse/DEVPROD-3142) to simplify the "who should review this" question with a `CODEOWNERS` file, once I get a little time to look at code quality infrastructure.

jimoleary reviewed (commented):
I haven't started looking at the actual workloads yet but I'm concerned about everything in the templates directory.

Do we really want this in the repo? And if we are going to add template generation to genny I don't think that the ```patch``` command is how it should be implemented.

jimoleary reviewed (commented):
I've added some general comments to the yml.base file but overall the workload looks ok.

Do you have any idea how noisy the new workload is? We would generally run a new workload 3 to 5 times to determine how noisy it is.

brettnawrocki commented on src/workloads/sharding/templates/multi_updates/MultiUpdates.yml.base:
```diff
@@ -0,0 +1,78 @@
+SchemaVersion: 2018-07-01
+Owner: "@mongodb/sharding"
+Description: |
+  This workload does the following:
```
Done.

brettnawrocki commented on src/workloads/sharding/templates/multi_updates/MultiUpdates.yml.base:
```diff
@@ -0,0 +1,78 @@
+SchemaVersion: 2018-07-01
+Owner: "@mongodb/sharding"
+Description: |
```
Done.

brettnawrocki commented on src/workloads/sharding/templates/multi_updates/MultiUpdates.yml.base:
```diff
@@ -0,0 +1,78 @@
+    Run updateMany as many times as possible for 5 minutes.
+    Each updateMany command will update only a single document.
+    The PauseMigrationsDuringMultiUpdates cluster parameter is disabled.
+
```
Done.

brettnawrocki commented on src/workloads/sharding/templates/multi_updates/MultiUpdates.yml.base:
```diff
@@ -0,0 +1,78 @@
+  Collection: &Collection coll
+  Namespace: &Namespace test.coll
+
+  ApproxDocumentSize: &ApproxDocumentSize 20_000
```
Done.

brettnawrocki commented on src/workloads/sharding/templates/multi_updates/MultiUpdates.yml.base:
```diff
@@ -0,0 +1,78 @@
+    Operations:
+    - OperationName: updateMany
+      OperationCommand:
+        Filter: { _id: 0 }
```
I took the first suggestion and did both kinds of updateManys in separate phases per workload. It does make sense that it's better to reuse the setup if we can for both variants. I didn't add the comments since I think it's clear enough now the purpose of each filter based on the name of the metrics.

brettnawrocki commented:
>I haven't started looking at the actual workloads yet but I'm concerned about everything in the templates directory.
>Do we really want this in the repo? And if we are going to add template generation to genny I don't think that the patch command is how it should be implemented.
>And I should add for full transparancy that I added some templating support in the past [here](https://github.com/mongodb/genny/tree/master/src/workloads/transactions/templates).

I expected this to be somewhat controversial, though as you mention there is some existing precedent for "bring your own code generation" for other workloads in the repo. I see [DEVPROD-3567](https://jira.mongodb.org/browse/DEVPROD-3567) was filed partially in reaction to this PR. I agree that using `patch` is not ideal and probably shouldn't be standardized, but for smaller scale usage like this, I think it has the advantage of being flexible and already well understood by the average engineer.

My hope is that a better alternative can be proposed (and implemented if necessary), but that this can be used in the meantime.

>Do you have any idea how noisy the new workload is? We would generally run a new workload 3 to 5 times to determine how noisy it is.

The cluster parameter in this workload isn't implemented yet (so enabling it does literally nothing). From the first patch, each workload is run effectively twice and I saw ~5% difference between them.

I planned to run [another patch](https://spruce.mongodb.com/version/659f1fc461837defe28f1979/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC) and then rerun the workload a few times to get better numbers, but my patch failed I believe due to [BF-31402](https://jira.mongodb.org/browse/BF-31402).

jimoleary reviewed (commented):
added a suggestion about parameterization rather than code generation.

brettnawrocki commented:
New patch build: https://spruce.mongodb.com/version/65a85231a4cf47ebaffecbdd/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC
I also ran an additional patch to get a larger sample size: https://spruce.mongodb.com/version/65a9504b5623434a2afaad70/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

**Results on Linux Shard Lite (all feature flags) 2022-11**
| | Unsharded, Update One | Unsharded, Update All | Sharded, Update One | Sharded, Update All |
| --- | --- | --- | --- | --- |
| Trial 1|130502 | 125 | 128031 | 201 |
| Trial 2|124107 | 117 | 130082 | 198 |
| Trial 3|132235 | 118 | 126664 | 198 |
| Trial 4|128496 | 121 | 118581 | 195 |
| Min |124107 | 117 | 118581 | 195 |
| Max |132235 | 125 | 130082 | 201 |
| Diff | 6.55% | 6.84% | 9.70% | 3.08% |

brettnawrocki commented on src/workloads/sharding/templates/multi_updates/MultiUpdates.yml.base:
```diff
@@ -15,7 +26,7 @@ GlobalDefaults:
   Namespace: &Namespace test.coll
 
-  ApproxDocumentSize: &ApproxDocumentSize 20_000
+  DocumentSize: &DocumentSize 20_000
```
Done.

jimoleary commented on src/phases/sharding/multi_updates/MultiUpdatesTemplate.yml:
```diff
@@ -0,0 +1,122 @@
+      updateMany will update all of the 50k documents per command in the first
+      phase and only a single document per command in the second phase.
+
+  Two parameters are accepted to alter the behavior of this workload:
```
Thanks for adding the documentation!

jimoleary reviewed (approved):
Thanks for the major refactor. 

I hope it wasn't too onerous but the workload looks much cleaner now.

brettnawrocki commented:
> Thanks for the major refactor.
> 
> I hope it wasn't too onerous but the workload looks much cleaner now.

It was no problem, and as I mentioned before, I expected the code generator to be controversial. 😅 I think the end result is definitely more maintainable than the original iteration.

The way `LoadConfig` and `OnlyActiveInPhases` can be combined to run some blocks conditionally was not obvious to me, so thanks for pointing that out. I do kind of wish there was some special value for `NopInPhasesUpTo` that meant "all other phases" instead of needing to pass that in (and potentially modify it in a number of places if another phase is added). Similarly, I wish I could have made the parameters e.g. `shardCollection: true|false` instead of needing to pass in `[]` or `[1]` to indicate the phases active, which again could change if the structure of the workload was updated.

alicedoherty commented:
Hey @brettnawrocki, just taking over from Jim on behalf of the performance team. We’d definitely encourage you to open a DEVPROD ticket (with the Assigned Teams field set to “Performance Infrastructure”) if you think having an “all other phases” feature in Genny would be valuable. Also, we have an [OnlyRunInInstance: sharded](https://github.com/mongodb/genny/blob/3512efc9284fc8f8928da02a8ea05f5386de35e4/src/workloads/execution/CreateBigIndex.yml#L27) feature that can be used to conditionally shard a collection (the downside of that is until [DEVPROD-2049](https://jira.mongodb.org/browse/DEVPROD-2049) is resolved you’ll need to exclude your workloads from dry-runs [here](https://github.com/mongodb/genny/blob/master/src/lamplib/src/genny/tasks/dry_run.py#L23-L39)).

None of this is blocking and I see Jim’s already approved this PR, but just adding this comment in case it adds any value!
